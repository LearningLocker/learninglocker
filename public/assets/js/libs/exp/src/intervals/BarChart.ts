/// <reference path="../definitions/references.d.ts" />
import React = require('react');
import Chart = require('react-chartjs');
var BarChart = Chart.Bar;
var DOM = React.DOM;

var TRANSPARENT_ORANGE = 'rgba(245,171,53,0.2)';
var OPAQUE_ORANGE = 'rgba(245,171,53,1)';
var OPAQUE_DARK_ORANGE = '#f4783b';
var DAYS = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
var MONTHS = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];

interface Props {
  data: Array<any>,
  interval: string
}

class Component extends React.Component<Props, any> {
  shouldComponentUpdate(next_props: Props): boolean {
    return this.props.data !== next_props.data;
  }
  render() {
    var graph_data = {
      labels: this.props.data.map(function (result) {
        switch (this.props.interval) {
          case 'Day-of-week': return DAYS[result._id - 1];
          case 'Month-of-year': return MONTHS[result._id - 1];
          case 'Results-tens': return '' + (result._id * 10) + '-' + ((result._id + 1) * 10) + '%';
          default: return result._id;
        }
      }.bind(this)),
      datasets: [{
        label: "Number of statements",
        fillColor: TRANSPARENT_ORANGE,
        strokeColor: OPAQUE_ORANGE,
        pointColor: OPAQUE_ORANGE,
        pointStrokeColor: OPAQUE_DARK_ORANGE,
        pointHighlightFill: "#fff",
        pointHighlightStroke: OPAQUE_DARK_ORANGE,
        data: this.props.data.map(function (result) {
          return result.count;
        })
      }]
    };

    var graph = this.props.data.length === 0 ? 'No data.' : React.createElement(BarChart, {
      data: graph_data,
      width: '600',
      height: '250',
      redraw: true // https://github.com/jhudson8/react-chartjs/issues/15
    });

    return DOM.div({}, [graph]);
  }
}

export = React.createFactory(Component);