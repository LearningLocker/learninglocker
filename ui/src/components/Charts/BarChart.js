import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { List } from 'immutable';
import { AutoSizer } from 'react-virtualized';
import { BarChart as Chart, XAxis, YAxis, CartesianGrid } from 'recharts';
import { compose } from 'recompose';
import NoData from 'ui/components/Graphs/NoData';
import { wrapLabel } from 'ui/utils/defaultTitles';
import { Button } from 'react-toolbox/lib/button';
import uuid from 'uuid';
import { connect } from 'react-redux';
import { getMetadataSelector, setInMetadata } from 'ui/redux/modules/metadata';
import {
  Buttons,
  Chart as StyledChart,
  ChartWrapper,
  BarContainer,
  XAxisLabel,
  XAxis as StyledXAxis,
  YAxis as StyledYAxis,
} from 'ui/components/Charts/styled';
import {
  getResultsData,
  getShortModel,
  getChartData,
  hasData,
  renderTooltips,
  renderBars,
  renderLegend,
  hiddenSeriesState
} from './Chart';

const enhance = compose(
  hiddenSeriesState,
  connect((state, { model }) =>
    ({
      activePage: getMetadataSelector({
        schema: 'visualisation',
        id: model.get('_id')
      })(state).get('activePage', 0)
    }), { setInMetadata })
);

class BarChart extends Component {
  static propTypes = {
    results: PropTypes.instanceOf(List),
    labels: PropTypes.instanceOf(List),
    colors: PropTypes.instanceOf(List),
    stacked: PropTypes.bool,
    chartWrapperFn: PropTypes.instanceOf(Function),
  }

  static defaultProps = {
    chartWrapperFn: component => (<AutoSizer>{component}</AutoSizer>)
  }

  displayPrevPage = () => this.props.setInMetadata({
    schema: 'visualisation',
    id: this.props.model.get('_id'),
    path: ['activePage'],
    value: this.props.activePage - 1
  });

  displayNextPage = () => this.props.setInMetadata({
    schema: 'visualisation',
    id: this.props.model.get('_id'),
    path: ['activePage'],
    value: this.props.activePage + 1
  });

  getDataChunk = model => data => page => data.slice(model.get('barChartGroupingLimit') * page, model.get('barChartGroupingLimit') * (page + 1))
  getPages = (model, data) => Math.ceil(data.size / (model.get('barChartGroupingLimit') + 1));
  hasPrevPage = pages => page => pages > 0 && page > 0
  hasNextPage = pages => page => pages > 0 && page < pages - 1
  sortData = data => data.sortBy(e => -e.get('total'))
  getSortedData = results => labels => (
    this.sortData(getResultsData(results)(labels))
  )

  renderPrevButton = () => (
    <span style={{ alignSelf: 'flex-start', marginLeft: 10 }}>
      <Button
        raised
        label="Previous"
        onMouseUp={this.displayPrevPage}
        icon={<i className="icon ion-chevron-left" />} />
    </span>
  )

  renderNextButton = () => (
    <span style={{ marginRight: 10, marginLeft: 'auto' }}>
      <Button
        raised
        label="Next"
        onMouseUp={this.displayNextPage}
        icon={<i className="icon ion-chevron-right" />} />
    </span>
  )

  renderBarChart = model => colors => labels => data => stacked => page => ({ width, height }) => {
    const chartUuid = uuid.v4();
    /* eslint-disable react/no-danger */
    return (
      <div>
        <style
          dangerouslySetInnerHTML={{
            __html: `
            .grid-${chartUuid} .recharts-cartesian-grid-horizontal {
              background-color: 'yellow';
              visibility: hidden !important;
            }
          ` }} />
        <Chart
          className={`grid-${chartUuid}`}
          data={getChartData(this.getDataChunk(model)(data)(page), this.props.hiddenSeries)}
          width={width}
          height={height}
          layout="vertical">
          <CartesianGrid strokeDasharray="1 1" />
          <YAxis
            dataKey="cellId"
            tickFormatter={getShortModel(data)}
            type="category"
            width={90} />
          <XAxis type="number" />
          {renderLegend(labels, this.props.toggleHiddenSeries)}
          {renderBars(colors)(labels)(stacked)}
          {renderTooltips(data, this.props.hiddenSeries)}
        </Chart>
      </div>
    );
    /* eslint-enable react/no-danger */
  };

  renderResults = model => results => colors => labels => (stacked) => {
    const { activePage } = this.props;
    const data = this.getSortedData(results)(labels);
    const pages = this.getPages(model, data);

    return (
      <StyledChart>
        <Buttons>
          {this.hasPrevPage(pages)(activePage) && this.renderPrevButton()}
          {this.hasNextPage(pages)(activePage) && this.renderNextButton()}
        </Buttons>
        <div className={'clearfix'} />
        <BarContainer>
          <StyledYAxis>
            {wrapLabel(this.props.model.get('axesyLabel') || this.props.model.getIn(['axesgroup', 'searchString'], 'Y-Axis'))}
          </StyledYAxis>
          <ChartWrapper>
            {this.props.chartWrapperFn((this.renderBarChart(model)(colors)(labels)(data)(stacked)(activePage)))}
          </ChartWrapper>
        </BarContainer>
        <XAxisLabel>
          <StyledXAxis>
            {this.props.model.get('axesxLabel') || this.props.model.getIn(['axesvalue', 'searchString'], 'X-Axis')}
          </StyledXAxis>
        </XAxisLabel>
      </StyledChart>
    );
  }

  render = () => {
    const { results, labels, stacked, colors, model } = this.props;
    return (
      hasData(this.props.results)
        ? this.renderResults(model)(results)(colors)(labels)(stacked)
        : <NoData />
    );
  }
}

export default enhance(BarChart);
