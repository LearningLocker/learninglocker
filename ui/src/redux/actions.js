export {
  decodeLoginTokenAction
} from 'ui/redux/modules/auth';
export {
  addModel,
  saveModel,
  deleteModel,
  expandModel,
  modelsWorker,
  fetchModel,
  updateModel,
  pollWhile
} from 'ui/redux/modules/models';
export {
  clearModelsCache,
  fetchModels,
  fetchModelsCount,
  fetchMore,
  fetchAllOutstandingModels
} from 'ui/redux/modules/pagination';
export {
  updateStatementQuery,
  updateStatementTimezone
} from 'ui/redux/modules/statements';
export {
  downloadExport
} from 'ui/redux/modules/exports';
export {
  deleteUserOrganisation,
} from 'ui/redux/modules/userOrganisations';
export {
  updateUserOrganisationSetting,
} from 'ui/redux/modules/userOrganisationSettings';
