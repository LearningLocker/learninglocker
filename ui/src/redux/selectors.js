export {
  modelsSelector,
  modelsSchemaSelector,
  modelsSchemaIdSelector,
  modelsPickSelector,
  modelsByFilterSelector,
  isLoadingModelSelector,
  shouldFetchModelSelector,
} from 'ui/redux/modules/models';
export {
  idsByFilterSelector,
  isLoadingSelector,
  isLoadingCountSelector,
  countSelector,
  hasMoreSelector,
  shouldFetchSelector,
  shouldFetchCountSelector,
  cursorSelector
} from 'ui/redux/modules/pagination';
export {
  statementQuerySelector,
  statementTimezoneSelector
} from 'ui/redux/modules/statements';
export {
  isAuthenticatedSelector,
  isLoggingOutSelector,
  isAuthenticatedWithOrgSelector,
  loggedInUserId,
  loggedInUserSelector,
  currentScopesSelector,
  organisationSettingsSelector,
  activeOrganisationSettingsSelector,
  isSiteAdminSelector
} from 'ui/redux/modules/auth';
