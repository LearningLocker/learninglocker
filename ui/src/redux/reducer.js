import { combineReducers } from 'redux';
import recycleState from 'redux-recycle';
import { router5Reducer } from 'redux-router5';

import auth, { LOGOUT, ORG_LOGOUT } from 'ui/redux/modules/auth';
import pagination from 'ui/redux/modules/pagination';
import models from 'ui/redux/modules/models';
import aggregation from 'ui/redux/modules/aggregation';
import statements from 'ui/redux/modules/statements';
import search from 'ui/redux/modules/search';
import uploadpeople from 'ui/redux/modules/people';
import exportsReducer from 'ui/redux/modules/exports';
import logo from 'ui/redux/modules/logo';
import clientIds from 'ui/redux/modules/clientIds';
import toasts from 'ui/redux/modules/toasts';
import queries from 'ui/redux/modules/queries';
import metadata from 'ui/redux/modules/metadata';
import app from 'ui/redux/modules/app';
import alerts from 'ui/redux/modules/alerts';

export default combineReducers({
  auth: recycleState(auth, [LOGOUT]),
  pagination: recycleState(pagination, [LOGOUT, ORG_LOGOUT]),
  aggregation: recycleState(aggregation, [LOGOUT, ORG_LOGOUT]),
  models: recycleState(models, [LOGOUT, ORG_LOGOUT]),
  uploadpeople: recycleState(uploadpeople, [LOGOUT, ORG_LOGOUT]),
  exports: recycleState(exportsReducer, [LOGOUT, ORG_LOGOUT]),
  logo: recycleState(logo, [LOGOUT, ORG_LOGOUT]),
  statements: recycleState(statements, [LOGOUT, ORG_LOGOUT]),
  search: recycleState(search, [LOGOUT, ORG_LOGOUT]),
  toasts: recycleState(toasts, [LOGOUT, ORG_LOGOUT]),
  clientIds: recycleState(clientIds, [LOGOUT, ORG_LOGOUT]),
  queries: recycleState(queries, [LOGOUT, ORG_LOGOUT]),
  metadata: recycleState(metadata, [LOGOUT, ORG_LOGOUT]),
  alerts: recycleState(alerts, [LOGOUT, ORG_LOGOUT]),
  app,
  router: router5Reducer
});
