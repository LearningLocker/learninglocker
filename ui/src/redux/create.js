import { createStore as _createStore, applyMiddleware, compose } from 'redux';
import { router5Middleware } from 'redux-router5';
import clientMiddleware from 'ui/redux/middleware/clientMiddleware';
import thunkMiddleware from 'redux-thunk';
import userMiddleware from 'ui/redux/middleware/userMiddleware';
import rootSaga, { sagaMiddleware } from 'ui/redux/modules/sagas';
import reducer from 'ui/redux/reducer';


export default function createStore(llClient, router, data) {
  const middleware = [
    thunkMiddleware,
    userMiddleware,
    clientMiddleware(llClient),
    sagaMiddleware,
    router5Middleware(router),
  ];

  const composeEnhancers =
    (typeof window === 'object' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__)
    ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    : compose;

  let finalCreateStore;
  if (__DEVELOPMENT__ && __CLIENT__ && __DEVTOOLS__) {
    const { persistState } = require('redux-devtools');

    finalCreateStore = composeEnhancers(
      applyMiddleware(...middleware),
      persistState(window.location.href.match(/[?&]debug_session=([^&]+)\b/))
    )(_createStore);
  } else {
    finalCreateStore = applyMiddleware(...middleware)(_createStore);
  }

  const store = finalCreateStore(reducer, data);
  sagaMiddleware.run(rootSaga);

  return store;
}

// UNICORN
