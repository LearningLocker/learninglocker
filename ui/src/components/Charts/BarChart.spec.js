import React from 'react';
import 'jest-styled-components';
import renderer from 'react-test-renderer';
import { List, fromJS, Map } from 'immutable';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import BarChart from './BarChart';

describe('BarChart', () => {
  test('should render', () => {
    const mockStore = createStore(
      (data, action) => {
        switch (action.type) {
          case '@@redux/INIT':
            return data;
          default:
            return data;
        }
      },
      {
        metadata: new Map()
      }
    );

    const mockModel = new Map({
      _id: '1111111111111111'
    });

    const barChart = renderer
      .create(
        <Provider store={mockStore}>
          <BarChart
            results={new List()}
            model={mockModel} />
        </Provider>
      )
      .toJSON();

    expect(barChart).toMatchSnapshot();
  });

  test('should render axes', () => {
    const mockResult = fromJS([[{
      '1970-01-01': {
        _id: '1970-01-01',
        count: 1,
        model: '1970-01-01'
      }
    }]]);

    const mockStore = createStore(
      (data, action) => {
        switch (action.type) {
          case '@@redux/INIT':
            return data;
          default:
            return data;
        }
      },
      { // state
        metadata: new Map()
      }
    );

    const mockModel = new Map({
      _id: '1111111111111111',
    });

    const barChart = renderer
      .create(
        <Provider store={mockStore}>
          <BarChart
            results={mockResult}
            labels={fromJS([undefined])}
            axesLabels={{
              yLabel: 'THE-Y',
              xLabel: 'THE-X'
            }}
            chartWrapperFn={() => (<div />)}
            model={mockModel} />
        </Provider>
      )
      .toJSON();

    expect(barChart).toMatchSnapshot();
  });
});
