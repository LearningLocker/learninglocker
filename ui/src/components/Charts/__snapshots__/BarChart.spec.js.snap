// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BarChart should render 1`] = `
<h4
  style={
    Object {
      "marginTop": 20,
      "textAlign": "center",
      "width": "100%",
    }
  }
>
  No data in this time range.
</h4>
`;

exports[`BarChart should render axes 1`] = `
.c0 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 svg g text {
  font-size: 12px;
}

.c4 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin-top: -4px;
}

.c6 {
  display: inline-block;
  white-space: nowrap;
}

.c5 {
  text-align: right;
  word-break: break-word;
  overflow: hidden;
  text-overflow: ellipsis;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: static;
  margin-right: 15px;
  margin-bottom: 10px;
}

.c3 {
  display: inline-block;
  -webkit-writing-mode: vertical-rl;
  -ms-writing-mode: tb-rl;
  writing-mode: vertical-rl;
  -webkit-transform: rotate(-180deg);
  -ms-transform: rotate(-180deg);
  transform: rotate(-180deg);
  margin: auto 0;
  white-space: pre-wrap;
}

<div
  className="c0"
>
  <div
    className="c1"
  />
  <div
    className="clearfix"
  />
  <div
    className="c2"
  >
    <div
      className="c3"
    >
      Y-Axis
    </div>
    <div
      className="c4"
    >
      <div />
    </div>
  </div>
  <div
    className="c5"
  >
    <span
      className="c6"
    >
      X-Axis
    </span>
  </div>
</div>
`;
